File: app/src/main/java/io/github/zwieback/familyfinance/business/article/activity/ArticleActivity.java
package io.github.zwieback.familyfinance.business.article.activity;

import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;

import com.annimon.stream.Objects;
import com.jakewharton.rxbinding3.appcompat.RxSearchView;

import java.util.concurrent.TimeUnit;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.business.article.filter.ArticleFilter;
import io.github.zwieback.familyfinance.business.article.lifecycle.destroyer.ArticleAsParentDestroyer;
import io.github.zwieback.familyfinance.business.article.lifecycle.destroyer.ArticleFromExpenseOperationsDestroyer;
import io.github.zwieback.familyfinance.business.article.listener.OnArticleClickListener;
import io.github.zwieback.familyfinance.core.activity.EntityFolderActivity;
import io.github.zwieback.familyfinance.core.fragment.EntityFolderFragment;
import io.github.zwieback.familyfinance.core.lifecycle.destroyer.EntityDestroyer;
import io.github.zwieback.familyfinance.core.model.Article;
import io.github.zwieback.familyfinance.core.model.ArticleView;
import io.github.zwieback.familyfinance.databinding.ItemArticleBinding;

import static io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.RESULT_ARTICLE_ID;
import static io.github.zwieback.familyfinance.util.NumberUtils.UI_DEBOUNCE_TIMEOUT;

public abstract class ArticleActivity<
        FRAGMENT extends EntityFolderFragment<ArticleView, ArticleFilter, ItemArticleBinding, OnArticleClickListener, ?>,
        FILTER extends ArticleFilter>
        extends EntityFolderActivity<ArticleView, Article, ArticleFilter, FRAGMENT>
        implements OnArticleClickListener {

    private MenuItem searchItem;

    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        getMenuInflater().inflate(R.menu.menu_entity_search, menu);
        searchItem = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) searchItem.getActionView();
        setupSearchItem(searchItem, menu);
        setupSearchView(searchView);
        return super.onCreateOptionsMenu(menu);
    }

    private void setupSearchItem(@Nullable MenuItem searchItem, @NonNull Menu menu) {
        if (searchItem == null) {
            return;
        }
        searchItem.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
            @Override
            public boolean onMenuItemActionExpand(MenuItem item) {
                showMenuGroupOfAddEntries(menu, false);
                return true;
            }

            @Override
            public boolean onMenuItemActionCollapse(MenuItem item) {
                showMenuGroupOfAddEntries(menu, true);
                // need invalidate options menu, because of platform bug:
                // see similar example: https://github.com/jakewharton/actionbarsherlock/issues/467
                invalidateOptionsMenu();
                return true;
            }
        });
    }

    private void setupSearchView(@NonNull SearchView searchView) {
        addDisposable(
                RxSearchView.queryTextChanges(searchView)
                        .debounce(UI_DEBOUNCE_TIMEOUT, TimeUnit.MILLISECONDS)
                        .subscribe(searchName -> {
                            getFilter().setSearchName(searchName.toString());
                            applyFilter(getFilter());
                        })
        );
    }

    @Nullable
    protected Integer getInitialParentId() {
        return getParentFolderId() != null ? getParentFolderId() : getDefaultParentId();
    }

    @Nullable
    abstract Integer getDefaultParentId();

    @NonNull
    @Override
    protected String getResultName() {
        return RESULT_ARTICLE_ID;
    }

    @Override
    protected boolean isFirstFrame() {
        return Objects.equals(getFilter().getParentId(), getInitialParentId());
    }

    @NonNull
    @Override
    protected String getFragmentTag() {
        return String.format("%s_%s", getLocalClassName(), getFilter().getParentId());
    }

    @Override
    public void onFolderClick(@NonNull View view, @NonNull ArticleView article) {
        closeSearchView();
        super.onFolderClick(view, article);
    }

    @Override
    protected void addEntity(int parentId, boolean isFolder) {
        closeSearchView();
        super.addEntity(parentId, isFolder);
    }

    @Override
    protected void editEntity(@NonNull ArticleView entity) {
        closeSearchView();
        super.editEntity(entity);
    }

    @NonNull
    @Override
    protected Class<Article> getClassOfRegularEntity() {
        return Article.class;
    }

    @NonNull
    @Override
    protected EntityDestroyer<Article> createDestroyer(ArticleView article) {
        if (article.isFolder()) {
            return new ArticleAsParentDestroyer(this, getData());
        }
        return new ArticleFromExpenseOperationsDestroyer(this, getData());
    }

    private void closeSearchView() {
        searchItem.collapseActionView();
    }

    private static void showMenuGroupOfAddEntries(Menu menu, boolean shown) {
        menu.setGroupEnabled(R.id.action_add_entries, shown);
        menu.setGroupVisible(R.id.action_add_entries, shown);
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/article/activity/ArticleActivity.kt
package io.github.zwieback.familyfinance.business.article.activity

import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.SearchView
import com.jakewharton.rxbinding3.appcompat.queryTextChanges
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.business.article.filter.ArticleFilter
import io.github.zwieback.familyfinance.business.article.lifecycle.destroyer.ArticleAsParentDestroyer
import io.github.zwieback.familyfinance.business.article.lifecycle.destroyer.ArticleFromExpenseOperationsDestroyer
import io.github.zwieback.familyfinance.business.article.listener.OnArticleClickListener
import io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.Companion.RESULT_ARTICLE_ID
import io.github.zwieback.familyfinance.constant.UiConstants.UI_DEBOUNCE_TIMEOUT
import io.github.zwieback.familyfinance.core.activity.EntityFolderActivity
import io.github.zwieback.familyfinance.core.fragment.EntityFolderFragment
import io.github.zwieback.familyfinance.core.lifecycle.destroyer.EntityDestroyer
import io.github.zwieback.familyfinance.core.model.Article
import io.github.zwieback.familyfinance.core.model.ArticleView
import io.github.zwieback.familyfinance.databinding.ItemArticleBinding
import java.util.concurrent.TimeUnit

abstract class ArticleActivity<FRAGMENT, FILTER> :
    EntityFolderActivity<ArticleView, Article, ArticleFilter, FRAGMENT>(),
    OnArticleClickListener
        where FRAGMENT : EntityFolderFragment<ArticleView, ArticleFilter, ItemArticleBinding, OnArticleClickListener, *>,
              FILTER : ArticleFilter {

    private var searchItem: MenuItem? = null

    protected val initialParentId: Int?
        get() = parentFolderId ?: defaultParentId

    protected abstract val defaultParentId: Int?

    override val resultName: String
        get() = RESULT_ARTICLE_ID

    override val isFirstFrame: Boolean
        get() = filter.getParentId() == initialParentId

    override val fragmentTag: String
        get() = "${localClassName}_${filter.getParentId()}"

    override val classOfRegularEntity: Class<Article>
        get() = Article::class.java

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_entity_search, menu)
        searchItem = menu.findItem(R.id.action_search)
        (searchItem?.actionView as? SearchView?)?.let { searchView ->
            setupSearchItem(searchItem, menu)
            setupSearchView(searchView)
        }
        return super.onCreateOptionsMenu(menu)
    }

    private fun setupSearchItem(searchItem: MenuItem?, menu: Menu) {
        searchItem?.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                showMenuGroupOfAddEntries(menu, false)
                return true
            }

            override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                showMenuGroupOfAddEntries(menu, true)
                // need invalidate options menu, because of platform bug:
                // see similar example: https://github.com/jakewharton/actionbarsherlock/issues/467
                invalidateOptionsMenu()
                return true
            }
        })
    }

    private fun setupSearchView(searchView: SearchView) {
        addDisposable(
            searchView.queryTextChanges()
                .debounce(UI_DEBOUNCE_TIMEOUT, TimeUnit.MILLISECONDS)
                .subscribe { searchName ->
                    filter.searchName = searchName.toString()
                    applyFilter(filter)
                }
        )
    }

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun onFolderClick(view: View, article: ArticleView) {
        closeSearchView()
        super.onFolderClick(view, article)
    }

    override fun addEntity(parentId: Int, isFolder: Boolean) {
        closeSearchView()
        super.addEntity(parentId, isFolder)
    }

    override fun editEntity(entity: ArticleView) {
        closeSearchView()
        super.editEntity(entity)
    }

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun createDestroyer(article: ArticleView): EntityDestroyer<Article> {
        return if (article.isFolder) {
            ArticleAsParentDestroyer(this, data)
        } else {
            ArticleFromExpenseOperationsDestroyer(this, data)
        }
    }

    private fun closeSearchView() {
        searchItem?.collapseActionView()
    }

    private fun showMenuGroupOfAddEntries(menu: Menu, shown: Boolean) {
        menu.setGroupEnabled(R.id.action_add_entries, shown)
        menu.setGroupVisible(R.id.action_add_entries, shown)
    }
}
